### DISCUSSION TITLE
Selecting multiple store props & actions best practice

### DISCUSSION THREAD

**Discussion by dannobytes** [pmndrs/zustand#1916] [Q&A] [answered]
Though there are several discussion threads about the proper usage of store selectors, it's still largely unclear what the "best" practice is for retrieving store props, actions or derived/computed values are in a performant way. Some threads suggest memoization, while others say it's unnecessary. 

My question here is a simple one and is really all about optimization. I want to know which pattern is more performant, i.e. minimizes rerenders.

```ts
// Select multiple props within a single selector
const [count, color, updateCount] = useMyStore(s => [s.count, s.color, s.updateCount])
```

VS

```ts
// Select each individual prop separately
const count = useMyStore(s => s.count)
const color = useMyStore(s => s.color)
const updateColor = useMyStore(s => s.updateColor)
```

Pattern 1 is much more consolidated and convenient. But because the selector is returning a new Array each time, would this cause the React component to always interpret this hook as having a new value, thus triggering a rerender each time it's run?

**Comment by SheikhArman100** [comment]
Pattern 2 seems more appropriate in my opinion because it won't unnecessarily re-render the component repeatedly. On the other hand, you can use shallow in pattern 1
`const [count, color, updateCount] = useMyStore(s => [s.count, s.color, s.updateCount],shallow)`
Shallow can help you minimise re-renders. Before re-rendering the component again, it will compare it to the previous array or object. the component will only re-render when state.count or state.color changes

**Comment by samhirtarif** [comment, marked as answer]
The [docs](https://github.com/pmndrs/zustand#selecting-multiple-state-slices) explain really well on how to access multiple slices/props from the state

From the docs

```js
import { shallow } from 'zustand/shallow'

// Object pick, re-renders the component when either state.nuts or state.honey change
const { nuts, honey } = useBearStore(
  (state) => ({ nuts: state.nuts, honey: state.honey }),
  shallow
)

// Array pick, re-renders the component when either state.nuts or state.honey change
const [nuts, honey] = useBearStore(
  (state) => [state.nuts, state.honey],
  shallow
)

// Mapped picks, re-renders the component when state.treats changes in order, count or keys
const treats = useBearStore((state) => Object.keys(state.treats), shallow)
```

You can also provide your own comparison function in place of `shallow`

**Reply by dannobytes** [reply]
Oh, thank you @samhirtarif for the pointer to that Recipes section in the docs. That must have been more recently updated since I last checked and didn't see it.

Sounds like the pattern for destructuring multiple props out of the store requires the `shallow` arg to be set, which is the main thing I missed. Seems like this is the "answer" here. So I'll mark this thread as done.

On a side note, these kinds of details seems buried in the docs under the "Guides" or "Recipes" section and isn't easy to find.

At some point, I do think an "API Reference" section could be really helpful. Often, devs who are already familiar with the overarching patterns/guides just need to see a list of all modules exported out of `zustand/*` like `createStore`, `create`, `shallow`, etc along with the various function signatures and how/when to use them.

**Comment by dannobytes** [comment]
Just a follow-up question here.

Is there any downside to always using the `shallow` equality function, even for single atomic state picks?

For example, is there any negative side effects to this? By default, changes uses strict === equality. But I'm wondering what would happen in this scenario.

```ts
import { shallow } from 'zustand/shallow';

const count = useShoeStore(s => s.count, shallow);
```

**Reply by dai-shi** [reply]
No, but it comes with extra tiny computation and extra bytes.

(fwiw, I'm considering removing `equalityFn` from the default hook toward zustand v5. There will be a migration path, but it will be out of standard. I will create an RFC.)

**Reply by dannobytes** [reply]
Thanks for the context. So it sounds like you're considering optimizing the equality function implicitly then? Or at least always defaulting to shallow?

Either way, it sounds like there's no downside in performance using the shallow comparison for even single picks but there is a downside if you don't include it for multiple state picks.

That helps to know! Basically, it lets me default to use shallow in my custom bound hook.

**Reply by dai-shi** [reply]
> So it sounds like you're considering optimizing the equality function implicitly then? Or at least always defaulting to shallow?

The other way around. The default behavior is strict equal `===`.

**Reply by dai-shi** [reply]
> Either way, it sounds like there's no downside in performance using the shallow comparison 

Yeah, shallow comparison is fairly cheap.
